src/conn/pipeline.rs

```diff
@@ -0,0 +1,80 @@
+use mysql_common::proto::codec::PacketCodecState;
+use std::{collections::VecDeque, marker::PhantomData};
+
+use crate::{
+    conn::{query_result::Binary, ConnMut},
+    prelude::*,
+    Conn, Params, QueryResult, Result, Statement,
+};
+
+struct PipelineInner<'c> {
+    conn: &'c mut Conn,
+    cmd_states: VecDeque<PacketCodecState>,
+}
+
+impl<'c> PipelineInner<'c> {
+    fn exec(&mut self, stmt: &Statement, params: Params) -> Result<()> {
+        self.conn._execute_pipeline(stmt, params)?;
+        self.cmd_states
+            .push_back(self.conn.stream_mut().codec().save_state());
+        Ok(())
+    }
+
+    fn finish(self) -> PipelineResult<'c, Binary> {
+        PipelineResult {
+            conn: self.conn,
+            cmd_states: self.cmd_states,
+            protocol: PhantomData,
+        }
+    }
+}
+
+pub struct Pipeline<'c>(Option<PipelineInner<'c>>);
+
+impl<'c> Pipeline<'c> {
+    pub(crate) fn new(conn: &'c mut Conn) -> Self {
+        Self(Some(PipelineInner {
+            conn,
+            cmd_states: VecDeque::new(),
+        }))
+    }
+
+    pub fn exec<P: Into<Params>>(&mut self, stmt: &Statement, params: P) -> Result<()> {
+        self.0.as_mut().unwrap().exec(stmt, params.into())
+    }
+
+    pub fn finish(mut self) -> PipelineResult<'c, Binary> {
+        self.0.take().unwrap().finish()
+    }
+}
+
+impl Drop for Pipeline<'_> {
```

Comment:
What happens if we don't implement `Drop` for pipeline or PipelineResult? Would this lead to a leak that we didn't reclaim the buffer for response packets?

Reply:
If we don't have a `Drop` on `PipelineResult`, then if you break out of the loop while reading query responses, the unread responses will screw up the next command you try to run on the connection. (This is the same reason `QueryResult` implements `Drop`.) The `should_pipeline_break_during_results_work` test covers this case.

If we don't have a `Drop` on `Pipeline`, the same thing happens if you drop the pipeline without calling `Finish`. Since the cleanup that needs to be done is basically the same as the first case, all it needs to do is (effectively) call `finish` itself, and then drop the `PipelineResult`.